### ISSUE TITLE
Slow concatenation

### ISSUE THREAD

**Issue by ivirshup** [theislab/anndata#303]
This seems far to slow for nothing to be going wrong. Might have to do with the overlapping obs_names?

```python
import scanpy as sc
from itertools import product

adata_ref = sc.datasets.pbmc3k()
```
```python
%%time
start = [596, 615, 1682, 1663, 1409, 1432]
adatas = [adata_ref[i : i + 1000] for i in start]
sample_names = [f"sa{i}_Rep{j}" for i, j in product((1, 2, 3), (1, 2))]

adata = sc.AnnData.concatenate(
    *adatas, join="outer", batch_key="sample_name", batch_categories=sample_names
)
adata.obs["time_points"] = adata.obs["sample_name"].str.split("_", expand=True)[0]
# CPU times: user 29.6 s, sys: 564 ms, total: 30.1 s
# Wall time: 30.4 s
```

Faster case:

```python
from functools import reduce

%time reduce(lambda x, y: x.concatenate(y), (adata_ref[i : i + 1000] for i in start))                               
# CPU times: user 1.15 s, sys: 141 ms, total: 1.29 s
Wall time: 1.29 s
```

**Comment by ivirshup** [comment]
Fixed by #352

First example now gives a result of:

```
CPU times: user 175 ms, sys: 35.4 ms, total: 210 ms
Wall time: 209 ms
```

**Comment by matthieuheitz** [comment]
For those arriving here because of slow concatenation, I had the same problem but it's not related to this bug.
I was subsetting the anndata by timepoint (I have ~150 of them), subsampling each timepoint, and then concatenating it back together.
Instead, I realized my adata.X was stored as a CSC matrix, which slowed down concatenation incredibly.
Instead, change it to CSR, and it will be super fast
`adata.X = adata.X.tocsr()`
